### PR TITLE
Disable LogColor on s390[x] LPARs and z/VMs.

### PR DESCRIPTION
**Pull request by xnox**
Fixes: https://github.com/systemd/systemd/issues/3024

### CODE REVIEW

**Comment by xnox** [comment]
Hehe, thanks @martinpitt that I don't have to test build this on s390x =)

**Comment by xnox** [comment]
Awwww my eyes, they don't hurt any more with this patch applied =) please please please merge this or something to the similar net effect (e.g. looking at the code potentially these checks can be done earlier too)

File: src/core/main.c [hunk 1]
```diff
@@ -1628,6 +1628,13 @@ int main(int argc, char *argv[]) {
                 if (v > 0)
                         log_info("Detected virtualization %s.", virtualization_to_string(v));
 
+#if defined(__s390__)
+                if (v == VIRTUALIZATION_NONE || v == VIRTUALIZATION_ZVM) {
+                        log_info("Disabling LogColor due to detected virtualization mode.");
```
Comment (keszybz) [inline src/core/main.c:1633]: log_debug would be better

**Comment by keszybz** [comment]
So... having arch specific code is not great, but OTOH, this patch is minimal. I'm +1 for merging, let's see if @poettering objects.

**Comment by poettering** [comment]
I am really not convinced this patch is a good idea, since it mixes up two unrelated things: which kind of TTY is connected to PID 1 and the CPU architecture. Now, of course, a specific type of TTY is often pretty commonly used with a specific CPU architecture, but there's no strict connection there. Especially in a world where you can run PID 1 easily in an nspawn container in an ssh session it sounds strange to connect this so strictly, because in that case the color output is generally of no bother...

And in general, I don't like arch-specific code in main.c I must say..

One option to deal with this might be to add a build-time configure option (`--disable-log-color`) that allows selecting the default, compiled-in use of colors. Then, when building systemd on s390 you can turn off color altogether. It's still broken to do that then, but at least that broken logic isn't in our code then, but a hack you do downstream in your build scripts.

Does that make sense?

**Comment by martinpitt** [comment]
Alternatively, could this only be done on `VIRTUALIZATION_ZVM` without the arch-specific `#ifdef`?

**Comment by keszybz** [comment]
> Alternatively, could this only be done on VIRTUALIZATION_ZVM without the arch-specific #ifdef?

According to the patch (and the comments in #3024), this isn't enough: you also want VIRTUALIZATION_NO on s390x to be covered.

> One option to deal with this might be to add a build-time configure option (--disable-log-color)

I think this will be actually more code (at least more files to change). The advantage of this solution is that it just works, but a separate configuration options is something that'll require a patch for the spec file, etc. It'll also cover more cases than necessary.

**Comment by martinpitt** [comment]
@xnox said that the usual console on these is some Java web app. Shouldn't these set a "poor" $TERM such as `vt100` or so? I guess the cleanest way of determining whether or not to display colors, aside from `isatty()` is which capabilities `$TERM` has? Although I don't know off-hand how to map a value to "can show colors".

@xnox, OOI, what is `$TERM` in these consoles?

**Comment by keszybz** [comment]
I thought we don't get `$TERM` for system consoles.

**Comment by keszybz** [comment]
... we get `$TERM` when running in a container, but not on "bare metal" or virtualization.

**Comment by poettering** [comment]
BTW, an earlier incarnation of this discussion was here:

https://lists.freedesktop.org/archives/systemd-devel/2014-August/021698.html

original mail:

https://lists.freedesktop.org/archives/systemd-devel/2014-June/020015.html

My proposal back then was this:

https://lists.freedesktop.org/archives/systemd-devel/2014-June/020339.html

And I think that's still very much valid. Except maybe that there is probably no need for a new setting, it can just be LogColor= as it already exists (note that strictly speaking the boot-time status output and the log output are different things, but I think it's fine binding the color use in the boot-time status output also to the log color use) . And of course my earlier suggestion in this PR to make the default for this build-time configurable applies still, too.

Hence maybe a solution like this:

a) bind the show status color use to LogColor=
b) make PID 1 look at TERM= on the kernel cmdline or in the env block passed to PID 1
c) allow a build-time default setting for LogColor=

(Note that the Linux kernel usually sets TERM=linux for PID 1 when invoking it with the normal console)

**Comment by keszybz** [comment]
> a) bind the show status color use to LogColor=

Hm, I thought we are already doing that since dc9b58166ca062583bcc8145ab3df79b2b39b9ae. Is there some place else which we should cover?

> b) make PID 1 look at TERM= on the kernel cmdline or in the env block passed to PID 1

Yep, we should do that.

> c) allow a build-time default setting for LogColor=

Is this still necessary when b) is implemented? What does the kernel set TERM= to on s390x?

**Comment by poettering** [comment]
Ah, nice, I missed that a) was actually already implemented. Sounds like with should then only do b) and maybe c).

**Comment by poettering** [comment]
With @keszybz's PR #3074 merged, both a) and b) are now implemented. @xnox could you report back if that solves your problem? or do we need c) still?

**Comment by martinpitt** [comment]
@xnox: Here, some s390x debs from upstream master (including the above mentioned changes) with Debian/Ubuntu patches applied. All tests pass on amd64, so the s390x ones should be good too: http://people.canonical.com/~pitti/tmp/systemd-master-20160420/

**Comment by xnox** [comment]
Tested b) specifying TERM=dumb on the kernel command line results in color-less output as expected. However, out of the box without any TERM variable on the command line pid1 environ gets TERM=linux by default. Even though in this case it clearly is not linux term.

Possibly instead of doing c) we could maybe make the default_term_for_tty to be smarted than just "linux or vt220", and have a case for dumb too?

**Comment by poettering** [comment]
@xnox oh, the default TERM=linux is passed to PID1 from the kernel really, it's not something PID1 itself sets for itself. See init/main.c in the kernel tree. I figure the kernel shouldn't set such rubbish if it isn't true on s390... My hopes in fixing the kernel to set something different there depending on whether a VC is actually configured as /dev/console are pretty small however.

Maybe indeed, systemd should actually override that early on, and use default_term_for_tty() for that. I'd probably even be fine to add an #ifdef s390 to that function to fix TERM to "dumb" if /dev/console is used (I am more open to #ifdeffery if its in library code and not in main code).

So, yeah, I'd be willing to merge two patches:

a) code that very early on in PID 1 "fixes" TERM via default_term_for_tty(), overriding what the kernel passed to us, but not overriding what is passed on the kernel cmdline

b) an update for default_term_for_tty() that returns TERM=dumb if /dev/console is tested on s390.

Does this make sense?

**Comment by xnox** [comment]
yeah, that makes sense. can we split that into a new issue? I might be able to look into it eventually.

**Comment by poettering** [comment]
@xnox sure, go ahead, file a new issue, and close #3025 and #3024.

**Comment by poettering** [comment]
OK, i did that now myself, new issue for this is #3115. Closing this one.

**Pull request closed without merging**